src/bytes.js

```diff
@@ -0,0 +1,1497 @@
+const supportedEncodings = {
+    "utf": "utf-8",
+    "utf8": "utf-8",
+    "utf-8": "utf-8",
+    "ascii": "ascii"
+};
+
+function normalizeEncoding(encoding) {
+    const normalized = encoding.replace(/\s+/g, "").toLowerCase();
+    const supported = supportedEncodings[normalized];
+    if (supported === undefined) {
+        return encoding;
+    } else {
+        return supported;
+    }
+}
+const Encoder = new TextEncoder();
+const Decoder = new TextDecoder();
+
+// Stop gap until Uint8Array.from (or new Uint8Array(iterable)) gets wider support
+// This only handles the simple case used in this file
+function Uint8ArrayFromArray(source) {
+    if (Uint8Array.from) {
+        return Uint8Array.from(source);
+    }
+
+    const uarr = new Uint8Array(source.length);
+
+    for (let idx = 0; idx < source.length; idx++) {
+        uarr[idx] = source[idx];
+    }
+
+    return uarr;
+}
+
+/**
+ * @constructor
+ * @param {*} source Using constructor with new should be a js object
+ * @param {Sk.builtin.str=} encoding Only called from python
+ * @param {Sk.builtin.str=} errors Only called from python
+ * @return {Sk.builtin.bytes}
+ * @extends {Sk.builtin.object}
+ */
+Sk.builtin.bytes = function (source, encoding, errors) {
+    if (!(this instanceof Sk.builtin.bytes)) {
+        // called from python
+        return newBytesFromPy(...arguments);
+    }
+
+    // deal with internal calls
+    if (source === undefined) {
+        this.v = new Uint8Array();
+    } else if (source instanceof Uint8Array) {
+        this.v = source;
+    } else if (Array.isArray(source)) {
+        Sk.asserts.assert(source.every((x) => x >= 0 && x < 256), "bad internal call to bytes with array");
+        this.v = Uint8ArrayFromArray(source);
+    } else if (typeof source === "string") {
+        this.v = Encoder.encode(source);
+    } else if (typeof source === "number") {
+        this.v = new Uint8Array(source);
+    } else {
+        // fall through case for subclassing called by Sk.abstr.superConstructor
+        const ret = Sk.misceval.chain(newBytesFromPy(...arguments), (pyBytes) => {
+            this.v = pyBytes.v;
+            return this;
+        });
+        // Sk.abstr.superConstructor is not suspension aware
+        return Sk.misceval.retryOptionalSuspensionOrThrow(ret);
+    }
+};
+
+Sk.abstr.setUpInheritance("bytes", Sk.builtin.bytes, Sk.builtin.seqtype);
+
+Sk.builtin.bytes.prototype.__class__ = Sk.builtin.bytes;
+
+function strEncode(pyStr, encoding, errors) {
+    const source = pyStr.$jsstr();
+    encoding = normalizeEncoding(encoding);
+    if (!(errors === "strict" || errors === "ignore" || errors === "replace")) {
+        throw new Sk.builtin.NotImplementedError("'" + errors + "' error handling not implemented in Skulpt");
+    }
+    let uint8;
+    if (encoding === "ascii") {
+        uint8 = encodeAscii(source, errors);
+    } else if (encoding === "utf-8") {
+        uint8 = Encoder.encode(source);
```

Comment:
Do we need to take account of `errors` here? Or will an encode always succeed into UTF-8, even if it's got weird codepoints in it? (There are some weird reserved ones like the UTF-16 surrogate range)

Reply:
@rixner - i pass this to you - equivalent line in #989 
https://github.com/skulpt/skulpt/blob/eba781b4ea810ee7100f18e0a7ac91dbedcfcc0e/src/bytes.js#L131

Reply:
@meredydd do you have an example string that you think might fail?  I am fairly certain that this is guaranteed to always encode into utf-8.  The input string is not being interpreted as utf-8.  The bytes constructor is encoding a sequence of codepoints U+0000 through U+00FF into utf-8.  So, it's not possible to have a "weird codepoint" in the input.

All of the errors will occur on decode.  I tried a bunch of potentially invalid utf sequences in both Python and TextEncoder and I get the same results from both every time. I couldn't produce an error in Python.  I tried invalid utf-8 bytes, utf-16 surrogate pairs, etc.  So I am pretty sure this can't fail.

Reply:
Wait, I think I'm wrong.  Looking into this further.

Reply:
Okay, I was wrong about how it was working, but I still believe it's not possible for this to fail. A javascript string can only contain valid unicode code points.  So, this just takes each code point and converts it into utf-8 bytes.  That should always work, unless there is an invalid code point in the input (I believe this to be impossible) or there is a code point that utf-8 can't represent (I believe that also cannot happen).

Reply:
This also seems how CPython is working.  So, I think they are compatible.  Anyone think I'm missing something?